### PR TITLE
Extend Model Builder to export Qwen 3.8 DFlash2 packages directly

### PR DESCRIPTION
**Pull request by baijumeswani**
## Description

This change extends the existing Qwen and DFlash2 Model Builder path so it can emit a complete Qwen 3.8 27B INT4 target and INT4 DFlash2 package directly from the original Hugging Face checkpoints. It removes the separate model-specific graph-surgery pipeline that previously transformed already-exported ONNX packages.

The implementation remains generic: target MLP gate/up fusion is an opt-in Model Builder optimization, and raw-layout CUDA `MatMulNBits` exports can explicitly select the fpA/intB kernel family. Existing DFlash2 support continues to own drafter construction, quantization, shared initializers, auxiliary taps, state wiring, serialization, and `genai_config.json`.

## Why

The repository already has the abstractions needed to build DFlash2 models. A second post-processing pipeline duplicated Model Builder responsibilities, required exact intermediate artifacts, and introduced another round of ONNX graph and external-data rewriting. Building the desired graph directly is simpler, easier to maintain, and uses the existing loader, quantizer, serializer, and test infrastructure.

## How the final model is built

The target comes from `Qwen/Qwen3.8-27B`, and the drafter comes from `incoai/Qwen3.8-27B-DFlash2`.

The Model Builder reads both checkpoints in one invocation. The target export enables paged attention, exposes the five SpecForge auxiliary streams at incoming layers `[6,20,34,48,62]`, emits compact state updates for seven draft tokens, retains full-row logits for block verification, quantizes target `MatMul` weights to symmetric INT4 with block size 32, and uses calibrated INT8 per-channel KV-cache scales.

`fuse_mlp_gate_up=true` combines each target MLP's gate and up projections before weight quantization, producing one `MatMulNBits` followed by `Split`. The existing DFlash2 builder quantizes eligible body weights to INT4, makes the drafter consume the target's exact quantized LM head, shares matching initializers with the target, and emits the DFlash2 section of `genai_config.json`.

`enable_cuda_fpa_intb_gemm=true` keeps the portable raw blockwise INT4 weight layout while writing `ep.cuda.fpa_intb_gemm=1` into the decoder session options. `use_device_allocator_for_initializers=true` avoids retaining replaced initializers in the ONNX Runtime arena during runtime prepacking.

```bash
PYTHONPATH=src/python/py python -m models.builder \
  -i /models/source/Qwen3.8-27B \
  -o /models/qwen3.8-27b-int4-int8-kv-dflash2 \
  -p int4 \
  -e cuda \
  --extra_options \
    exclude_embeds=false \
    use_paged_attention=true \
    paged_block_size=256 \
    max_batch_size=16 \
    gpu_utilization_factor=0.5 \
    enable_cuda_graph=false \
    prune_lm_head=false \
    block_size=32 \
    op_types_to_quantize=MatMul \
    linear_attn_op=gated_delta_net \
    state_update_capacity=7 \
    kv_cache_quant_scheme=int8_per_channel \
    kv_cache_scale_file=/models/qwen38-int8-kv-scales.json \
    aux_hidden_state_layers=6,20,34,48,62 \
    dflash2_path=/models/source/Qwen3.8-27B-DFlash2 \
    dflash2_num_draft_tokens=7 \
    dflash2_precision=int4 \
    fuse_mlp_gate_up=true \
    enable_cuda_fpa_intb_gemm=true \
    use_device_allocator_for_initializers=true
```

The output package contains `model.onnx`, `model.onnx.data`, `dflash2.onnx`, `dflash2.onnx.data`, `genai_config.json`, and the source checkpoint's tokenizer and processing files. No intermediate ONNX package or follow-up graph surgery is required.

## Changes

- Add opt-in target MLP gate/up projection fusion before target quantization.
- Add opt-in raw-layout CUDA fpA/intB session selection.
- Extend the existing Qwen DFlash2 tests for target fusion and package settings.
- Document both generic options and the direct DFlash2 build flow.
- Remove the superseded Qwen 3.8-specific post-processing package and tests.

## Correctness and validation

- The complete Model Builder suite passes: 725 tests passed with 4 environment-dependent skips.
- Python compilation, focused Ruff, and whitespace checks pass.
- The end-to-end direct source export completed from the original target and DFlash2 checkpoints.
- The target has the expected 136 inputs and 226 outputs, including all 96 compact state-update outputs and the five-layer auxiliary hidden-state output.
- All 916 target initializers with unchanged names, types, and shapes are byte-identical to the qualified package. Renamed Qwen linear-attention QKV and Z projection tensors are also byte-identical after name normalization.
- The DFlash2 input/output contract and all 141 initializers common to both packages are identical.
- The direct package shares the embedding and both target LM-head tensors and uses the qualified taps, draft width, block size, selector top-k, sliding window, and runtime session options.
- Deterministic greedy generation on an NVIDIA A100 produced exact token-for-token equality between the direct package with DFlash2 enabled and the same target with DFlash2 disabled for two prompts. The chat-formatted case also exactly matched the qualified package.
- Runtime telemetry confirmed real speculative execution across full, partial, and zero-acceptance rounds, with no DFlash2 failures, disables, or standard fallback steps.
- DFlash2 uses a dedicated session override for its raw BF16 body, validates external-data ranges before sharing, and keeps a private raw quantized LM head whenever the target prepacked layout cannot be adopted safely.
- Five independent reviews covering code, design, continuous-batching integration, generated-graph correctness, and runtime token correctness approved the implementation with no remaining blockers.

### CODE REVIEW

File: src/python/py/models/builders/base.py [hunk 1]
```diff
@@ -1361,9 +1363,11 @@ def make_genai_config(self, config, extra_kwargs, out_dir):
             genai_config["model"]["decoder"]["session_options"]["provider_options"].append(ep_options)
 
         session_options = genai_config["model"]["decoder"]["session_options"]
-        if self.ep == "cuda" and self.matmul_attrs["weights_prepacked"] > 0:
-            # Prepacked nodes take the fpA_intB path unconditionally; setting the flag keeps the
-            # nodes that were skipped (unsupported N/K/block_size) on the same kernel family.
+        if self.ep == "cuda" and (
+            self.matmul_attrs["weights_prepacked"] > 0 or self.extra_options.get("enable_cuda_fpa_intb_gemm", False)
+        ):
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/python/py/models/builders/base.py:1368]: Addressed. The DFlash2 section now emits its own `session_options` override with `ep.cuda.fpa_intb_gemm=0`, preventing the target decoder setting from being inherited by the raw BF16 drafter body while preserving provider options. `DrafterSessionOverridesTargetConfigEntries` verifies the final projected runtime config. The rebuilt DFlash2 C++ suite passes 52/52.

File: src/python/py/models/builders/mtp.py [hunk 2]
```diff
@@ -106,12 +113,8 @@ def find_shared_initializers(self, source_model, target_model, source_data, targ
...
-            if self.external_data_equal(
-                source_data,
-                source_offset,
-                target_data,
-                target_tensor.offset,
-                source_length,
+            if name in adopt_source_initializers or self.external_data_equal(
+                source_data, source_offset, target_data, target_tensor.offset, source_length
             ):
                 shared[name] = (source_offset, source_length, target_tensor.offset)
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/python/py/models/builders/mtp.py:126]: Addressed. Source external-data offsets and lengths are now validated against the source file size before either byte comparison or explicit adoption. A truncated adopted source-range regression verifies that sharing fails transactionally and leaves both auxiliary files unchanged.

File: src/python/py/models/builders/qwen.py [hunk 3]
```diff
@@ -1226,29 +1226,27 @@ def save_dflash2_model(self, output_dir):
         if self.dflash2 is None:
             return
         self.dflash2.save_model(output_dir)
+        adopted_head = set()
+        if getattr(self.dflash2, "lm_head_quant", None) is not None:
+            bits = self.dflash2.lm_head_quant["bits"]
+            adopted_head = {
+                f"lm_head.MatMul.weight_Q{bits}",
+                "lm_head.MatMul.weight_scales",
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/python/py/models/builders/qwen.py:1253]: Addressed using the effective layout contract. Non-CUDA targets normalize requested prepacking to raw layout. CUDA BF16 targets with prepacked weights now keep a private raw quantized drafter LM head (`adopt_target=False`) and exclude its weight and scales from ordinary sharing, preventing accidental raw/prepacked aliasing without falling back to a dense multi-gigabyte head. Compatible layouts still require all-or-nothing exact target adoption.

File: src/python/py/models/builders/base.py [hunk 4]
```diff
@@ -4895,6 +4905,98 @@ def make_mlp_proj(self, layer_id, mlp, root_input):
...
+        basename = f"/model/layers.{layer_id}/mlp"
+        gate_up_proj = types.SimpleNamespace(
+            weight=torch.cat((mlp.gate_proj.weight, mlp.up_proj.weight)),
+            bias=torch.cat((gate_bias, up_bias)) if bias_exists else None,
+            exclude_from_quantization=(
+                getattr(mlp.gate_proj, "exclude_from_quantization", False)
+                or getattr(mlp.up_proj, "exclude_from_quantization", False)
+            ),
+        )
+        matmul_name = self.make_matmul(gate_up_proj, f"{basename}/gate_up_proj/MatMul", root_input)
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/python/py/models/builders/base.py:4973]: Addressed. `linear_attn` mixed precision now targets `gate_up_proj` when fusion is enabled. Matching gate/up exclusions transfer to the fused node, while one-sided exclusions are rejected because they cannot be preserved by one MatMul. The Qwen MoE shared-expert path now also uses the fused builder and restores its temporary intermediate size with `finally`. Focused regressions cover each case.

**Review by tianleiwu — APPROVED** [review]
Reviewed the full origin/main...b8e8ec2d diff, including target gate/up fusion, DFlash2 initializer adoption and session-option isolation, quantizer routing, state wiring, and the associated regression coverage. I did not find any correctness issues that need changes. Validation: the five changed Python builder test modules pass locally (174 passed, 1 skipped).
